File: android/app/src/main/java/com/alphacity/stamptour/viewmodel/MyPageViewModel.kt
package com.alphacity.stamptour.viewmodel

import android.net.Uri
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.alphacity.stamptour.network.TokenManager
import com.alphacity.stamptour.network.dto.UserProfile
import com.alphacity.stamptour.repository.AuthRepository
import com.alphacity.stamptour.repository.HomeRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class MyPageViewModel @Inject constructor(
    private val homeRepository: HomeRepository,
    private val authRepository: AuthRepository,
    private val tokenManager: TokenManager,
) : ViewModel() {

    private val _userProfile = MutableStateFlow<UserProfile?>(null)
    val userProfile: StateFlow<UserProfile?> = _userProfile

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    private val _saveSuccess = MutableStateFlow(false)
    val saveSuccess: StateFlow<Boolean> = _saveSuccess

    private val _saveError = MutableStateFlow<String?>(null)
    val saveError: StateFlow<String?> = _saveError

    private val _deleteSuccess = MutableStateFlow(false)
    val deleteSuccess: StateFlow<Boolean> = _deleteSuccess

    private val _deleteError = MutableStateFlow<String?>(null)
    val deleteError: StateFlow<String?> = _deleteError

    // SMS 인증 상태
    private val _isCodeSent = MutableStateFlow(false)
    val isCodeSent: StateFlow<Boolean> = _isCodeSent

    private val _isPhoneVerified = MutableStateFlow(false)
    val isPhoneVerified: StateFlow<Boolean> = _isPhoneVerified

    private val _verificationError = MutableStateFlow<String?>(null)
    val verificationError: StateFlow<String?> = _verificationError

    // 로컬에서 선택한 프로필 이미지 URI (앱 세션 동안 유지)
    private val _profileImageUri = MutableStateFlow<Uri?>(null)
    val profileImageUri: StateFlow<Uri?> = _profileImageUri

    fun setProfileImageUri(uri: Uri?) {
        _profileImageUri.value = uri
    }

    val isLoggedIn: Boolean
        get() = tokenManager.isLoggedIn

    fun fetchProfile() {
        if (!tokenManager.isLoggedIn) return
        _isLoading.value = true

        viewModelScope.launch {
            homeRepository.getUserProfile()
                .onSuccess { _userProfile.value = it }
                .onFailure { Log.e("MyPageViewModel", "프로필 로드 실패", it) }
            _isLoading.value = false
        }
    }

    fun sendCode(phone: String) {
        if (phone.isBlank()) {
            _verificationError.value = "휴대폰 번호를 입력하세요"
            return
        }
        viewModelScope.launch {
            _isLoading.value = true
            _verificationError.value = null
            authRepository.sendCode(phone)
                .onSuccess {
                    _isCodeSent.value = true
                }
                .onFailure { e ->
                    _verificationError.value = e.message ?: "SMS 전송에 실패했습니다"
                }
            _isLoading.value = false
        }
    }

    fun verifyCode(phone: String, code: String) {
        if (code.isBlank()) {
            _verificationError.value = "인증번호를 입력하세요"
            return
        }
        viewModelScope.launch {
            _isLoading.value = true
            _verificationError.value = null
            authRepository.verifyCode(phone, code)
                .onSuccess {
                    _isPhoneVerified.value = true
                }
                .onFailure { e ->
                    _verificationError.value = e.message ?: "인증번호가 올바르지 않습니다"
                }
            _isLoading.value = false
        }
    }

    fun setPhoneVerifiedFromProfile() {
        _isPhoneVerified.value = true
    }

    fun resetVerificationState() {
        _isCodeSent.value = false
        _isPhoneVerified.value = false
        _verificationError.value = null
    }

    fun updateProfile(nickname: String, currentPassword: String?, newPassword: String?, phone: String? = null, name: String? = null, address: String? = null, addressDetail: String? = null, birthDate: String? = null, gender: String? = null) {
        _isLoading.value = true
        _saveError.value = null
        viewModelScope.launch {
            homeRepository.updateProfile(nickname, currentPassword, newPassword, phone, name, address, addressDetail, birthDate, gender)
                .onSuccess { updated ->
                    _userProfile.value = updated
                    _saveSuccess.value = true
                }
                .onFailure { _saveError.value = it.message ?: "저장에 실패했습니다." }
            _isLoading.value = false
        }
    }

    fun clearSaveState() {
        _saveSuccess.value = false
        _saveError.value = null
    }

    fun deleteAccount() {
        _isLoading.value = true
        _deleteError.value = null
        viewModelScope.launch {
            homeRepository.deleteAccount()
                .onSuccess {
                    tokenManager.clearTokens()
                    _userProfile.value = null
                    _deleteSuccess.value = true
                }
                .onFailure {
                    _deleteError.value = it.message ?: "회원 탈퇴에 실패했습니다."
                }
            _isLoading.value = false
        }
    }

    fun clearDeleteState() {
        _deleteSuccess.value = false
        _deleteError.value = null
    }

    fun logout() {
        tokenManager.clearTokens()
        _userProfile.value = null
    }
}

File: ios/AlphaCityStampTour/AlphaCityStampTour/ViewModel/MyPageViewModel.swift
//
//  MyPageViewModel.swift
//  AlphaCityStampTour
//

import Foundation

@MainActor
final class MyPageViewModel: ObservableObject {
    @Published var userProfile: UserProfileData?
    @Published var isLoading = false
    @Published var saveSuccess = false
    @Published var saveError: String?
    @Published var deleteSuccess = false
    @Published var deleteError: String?
    @Published var isCodeSent = false
    @Published var isPhoneVerified = false
    @Published var verificationError: String?

    private let repository = HomeRepository.shared
    private let tokenManager = TokenManager.shared
    private let client = APIClient.shared

    var isLoggedIn: Bool {
        tokenManager.isLoggedIn
    }

    var isSocialLogin: Bool {
        userProfile?.provider != nil
    }

    func fetchProfile() async {
        guard tokenManager.isLoggedIn else { return }
        isLoading = true
        do {
            userProfile = try await repository.fetchUserProfile()
        } catch {
            print("MyPageViewModel: 프로필 로드 실패 - \(error)")
        }
        isLoading = false
    }

    func sendCode(phone: String) {
        guard !phone.isEmpty else {
            verificationError = "휴대폰 번호를 입력하세요"
            return
        }
        isLoading = true
        verificationError = nil
        Task {
            do {
                let _: EmptyData? = try await client.request(
                    path: "auth/send-code",
                    method: "POST",
                    body: SendCodeRequest(phone: phone)
                )
                isCodeSent = true
            } catch {
                verificationError = error.localizedDescription
            }
            isLoading = false
        }
    }

    func verifyCode(phone: String, code: String) {
        guard !code.isEmpty else {
            verificationError = "인증번호를 입력하세요"
            return
        }
        isLoading = true
        verificationError = nil
        Task {
            do {
                let _: EmptyData? = try await client.request(
                    path: "auth/verify-code",
                    method: "POST",
                    body: VerifyCodeRequest(phone: phone, code: code)
                )
                isPhoneVerified = true
            } catch {
                verificationError = error.localizedDescription
            }
            isLoading = false
        }
    }

    func setPhoneVerifiedFromProfile() {
        isPhoneVerified = true
    }

    func resetVerificationState() {
        isCodeSent = false
        isPhoneVerified = false
        verificationError = nil
    }

    func updateProfile(nickname: String, currentPassword: String?, newPassword: String?, phone: String? = nil, name: String? = nil, address: String? = nil, addressDetail: String? = nil, birthDate: String? = nil, gender: String? = nil) async {
        isLoading = true
        saveError = nil
        do {
            let updated = try await repository.updateProfile(nickname: nickname, currentPassword: currentPassword, newPassword: newPassword, phone: phone, name: name, address: address, addressDetail: addressDetail, birthDate: birthDate, gender: gender)
            userProfile = updated
            saveSuccess = true
        } catch {
            saveError = error.localizedDescription
        }
        isLoading = false
    }

    func deleteAccount() async {
        isLoading = true
        deleteError = nil
        do {
            try await repository.deleteAccount()
            tokenManager.clearTokens()
            userProfile = nil
            deleteSuccess = true
        } catch {
            deleteError = error.localizedDescription
        }
        isLoading = false
    }

    func clearSaveState() {
        saveSuccess = false
        saveError = nil
    }

    func clearDeleteState() {
        deleteSuccess = false
        deleteError = nil
    }

    func logout() {
        tokenManager.clearTokens()
        userProfile = nil
    }
}
